### ISSUE TITLE
Cleanup on shutdown

### ISSUE THREAD

**Issue by lukasblu** [hazelcast/hazelcast#4354]
Hi,

we are using Hazelcast 3.4.1-SNAPSHOT as a library embedded in a web application running on Tomcat.

When Tomcat stops, it performs a few cleanup checks. There are two (non serious) issues with Hazelcast:

Here is our log output:

1st we perform an internal shutdown of the web application which also stops Hazelcast. Here all is fine.

```
2014-12-23 11:02:58,925 INFO  [Appway Shutdown Message Listener] com.hazelcast.core.LifecycleService - [10.0.1.104]:5713 [blade] [3.4.1-dev-20141222] Address[10.0.1.104]:5713 is SHUTTING_DOWN
2014-12-23 11:02:58,929 INFO  [Appway Shutdown Message Listener] com.hazelcast.instance.Node - [10.0.1.104]:5713 [blade] [3.4.1-dev-20141222] Shutting down connection manager...
2014-12-23 11:02:58,931 INFO  [Appway Shutdown Message Listener] com.hazelcast.instance.Node - [10.0.1.104]:5713 [blade] [3.4.1-dev-20141222] Shutting down node engine...
2014-12-23 11:02:59,004 INFO  [Appway Shutdown Message Listener] com.hazelcast.instance.NodeExtension - [10.0.1.104]:5713 [blade] [3.4.1-dev-20141222] Destroying node NodeExtension.
2014-12-23 11:02:59,004 INFO  [Appway Shutdown Message Listener] com.hazelcast.instance.Node - [10.0.1.104]:5713 [blade] [3.4.1-dev-20141222] Hazelcast Shutdown is completed in 76 ms.
2014-12-23 11:02:59,006 INFO  [Appway Shutdown Message Listener] com.hazelcast.core.LifecycleService - [10.0.1.104]:5713 [blade] [3.4.1-dev-20141222] Address[10.0.1.104]:5713 is SHUTDOWN
2014-12-23 11:02:59,006 INFO  [Appway Shutdown Message Listener] com.nm.cluster.service.ClusterServiceImpl - Hazelcast shutdown done (node detached).
2014-12-23 11:02:59,006 INFO  [Appway Shutdown Message Listener] com.nm.utils.SystemServicesManager - shutdown cluster service (hazelcast) done.
2014-12-23 11:02:59,006 INFO  [Appway Shutdown Message Listener] com.nm.utils.SystemServicesManager - shutdown lock pools...
2014-12-23 11:02:59,007 INFO  [Appway Shutdown Message Listener] com.nm.utils.SystemServicesManager - shutdown lock pools done.
2014-12-23 11:02:59,007 INFO  [Appway Shutdown Message Listener] com.nm.utils.SystemServicesManager - shutdown plugin extensions...
2014-12-23 11:02:59,007 INFO  [Appway Shutdown Message Listener] com.nm.services.AdapterServiceImpl - Stopping Extension: NmLoginServiceAdapter
2014-12-23 11:02:59,406 INFO  [Appway Shutdown Message Listener] com.nm.utils.SystemServicesManager - shutdown plugin extensions done.
2014-12-23 11:02:59,406 INFO  [Appway Shutdown Message Listener] com.nm.utils.SystemServicesManager - Appway will soon be stopped, but Application Server might continue running...
```

Then we stop the application server and here Tomcat checks for left-over thread-local data.

```
2014-12-23 11:03:15,155 INFO  [main] org.apache.coyote.http11.Http11Protocol - Pausing Coyote HTTP/1.1 on http-8080
2014-12-23 11:03:16,155 INFO  [main] org.apache.catalina.core.StandardService - Stopping service Catalina
2014-12-23 11:03:16,164 INFO  [main] org.apache.catalina.core.ContainerBase.[Catalina].[localhost].[/] -  [CompressingFilter/1.7.1] CompressingFilter is being destroyed...
2014-12-23 11:03:16,164 INFO  [main] org.apache.catalina.core.ContainerBase.[Catalina].[localhost].[/] -  [CompressingFilter/1.7.1] CompressingFilter is being destroyed...
2014-12-23 11:03:16,308 ERROR [main] org.apache.catalina.loader.WebappClassLoader - The web application [] appears to have started a thread named [Abandoned connection cleanup thread] but has failed to stop it. This is very likely to create a memory leak.
2014-12-23 11:03:16,310 ERROR [main] org.apache.catalina.loader.WebappClassLoader - The web application [] appears to have started a thread named [aw.event.user - 1] but has failed to stop it. This is very likely to create a memory leak.
2014-12-23 11:03:16,310 ERROR [main] org.apache.catalina.loader.WebappClassLoader - The web application [] created a ThreadLocal with key of type [org.apache.log4j.helpers.ThreadLocalMap] (value [org.apache.log4j.helpers.ThreadLocalMap@20390c93]) and a value of type [java.util.Hashtable] (value [{versionFilter=HeadFilter-permanent}]) but failed to remove it when the web application was stopped. This is very likely to create a memory leak.
2014-12-23 11:03:16,311 ERROR [main] org.apache.catalina.loader.WebappClassLoader - The web application [] created a ThreadLocal with key of type [com.hazelcast.util.UuidUtil$1] (value [com.hazelcast.util.UuidUtil$1@36a9618c]) and a value of type [java.util.Random] (value [java.util.Random@2a434a5d]) but failed to remove it when the web application was stopped. This is very likely to create a memory leak.
2014-12-23 11:03:16,311 ERROR [main] org.apache.catalina.loader.WebappClassLoader - The web application [] created a ThreadLocal with key of type [java.lang.ThreadLocal] (value [java.lang.ThreadLocal@27c73558]) and a value of type [com.nm.utils.ThreadInfo] (value [com.nm.utils.ThreadInfo@77f38cf]) but failed to remove it when the web application was stopped. This is very likely to create a memory leak.
2014-12-23 11:03:16,312 ERROR [main] org.apache.catalina.loader.WebappClassLoader - The web application [] created a ThreadLocal with key of type [com.hazelcast.util.UuidUtil$1] (value [com.hazelcast.util.UuidUtil$1@36a9618c]) and a value of type [java.util.Random] (value [java.util.Random@4fa9d41f]) but failed to remove it when the web application was stopped. This is very likely to create a memory leak.
2014-12-23 11:03:16,314 ERROR [main] org.apache.catalina.loader.WebappClassLoader - The web application [] created a ThreadLocal with key of type [com.hazelcast.util.UuidUtil$1] (value [com.hazelcast.util.UuidUtil$1@36a9618c]) and a value of type [java.util.Random] (value [java.util.Random@45544650]) but failed to remove it when the web application was stopped. This is very likely to create a memory leak.
2014-12-23 11:03:16,315 ERROR [main] org.apache.catalina.loader.WebappClassLoader - The web application [] created a ThreadLocal with key of type [com.hazelcast.util.UuidUtil$1] (value [com.hazelcast.util.UuidUtil$1@36a9618c]) and a value of type [java.util.Random] (value [java.util.Random@2c42e2ec]) but failed to remove it when the web application was stopped. This is very likely to create a memory leak.
(the same message repeated about 50 times more)
2014-12-23 11:03:16,363 ERROR [main] org.apache.catalina.loader.WebappClassLoader - The web application [] created a ThreadLocal with key of type [com.hazelcast.util.UuidUtil$1] (value [com.hazelcast.util.UuidUtil$1@36a9618c]) and a value of type [java.util.Random] (value [java.util.Random@56069c19]) but failed to remove it when the web application was stopped. This is very likely to create a memory leak.
2014-12-23 11:03:16,363 ERROR [main] org.apache.catalina.loader.WebappClassLoader - The web application [] created a ThreadLocal with key of type [com.hazelcast.util.UuidUtil$1] (value [com.hazelcast.util.UuidUtil$1@36a9618c]) and a value of type [java.util.Random] (value [java.util.Random@4bf2596e]) but failed to remove it when the web application was stopped. This is very likely to create a memory leak.
2014-12-23 11:03:16,364 ERROR [main] org.apache.catalina.loader.WebappClassLoader - The web application [] created a ThreadLocal with key of type [com.hazelcast.util.UuidUtil$1] (value [com.hazelcast.util.UuidUtil$1@36a9618c]) and a value of type [java.util.Random] (value [java.util.Random@13cd6bb6]) but failed to remove it when the web application was stopped. This is very likely to create a memory leak.
java.lang.NullPointerException
	at com.hazelcast.nio.ClassLoaderUtil$ConstructorCache.get(ClassLoaderUtil.java:159)
	at com.hazelcast.nio.ClassLoaderUtil.newInstance(ClassLoaderUtil.java:62)
	at com.hazelcast.logging.Logger.loadLoggerFactory(Logger.java:78)
	at com.hazelcast.logging.Logger.newLoggerFactory(Logger.java:57)
	at com.hazelcast.logging.Logger.getLogger(Logger.java:40)
	at com.hazelcast.logging.Logger.getLogger(Logger.java:30)
	at com.hazelcast.config.FileSystemXmlConfig.<clinit>(FileSystemXmlConfig.java:35)
	at sun.misc.Unsafe.ensureClassInitialized(Native Method)
	at sun.reflect.UnsafeFieldAccessorFactory.newFieldAccessor(UnsafeFieldAccessorFactory.java:43)
	at sun.reflect.ReflectionFactory.newFieldAccessor(ReflectionFactory.java:140)
	at java.lang.reflect.Field.acquireFieldAccessor(Field.java:1044)
	at java.lang.reflect.Field.getFieldAccessor(Field.java:1025)
	at java.lang.reflect.Field.get(Field.java:383)
	at org.apache.catalina.loader.WebappClassLoader.clearReferencesStaticFinal(WebappClassLoader.java:2066)
	at org.apache.catalina.loader.WebappClassLoader.clearReferences(WebappClassLoader.java:1929)
	at org.apache.catalina.loader.WebappClassLoader.stop(WebappClassLoader.java:1833)
	at org.apache.catalina.loader.WebappLoader.stop(WebappLoader.java:740)
	at org.apache.catalina.core.StandardContext.stop(StandardContext.java:4920)
	at org.apache.catalina.core.ContainerBase.stop(ContainerBase.java:1110)
	at org.apache.catalina.core.ContainerBase.stop(ContainerBase.java:1110)
	at org.apache.catalina.core.StandardEngine.stop(StandardEngine.java:468)
	at org.apache.catalina.core.StandardService.stop(StandardService.java:604)
	at org.apache.catalina.core.StandardServer.stop(StandardServer.java:788)
	at org.apache.catalina.startup.Catalina.stop(Catalina.java:662)
	at org.apache.catalina.startup.Catalina.start(Catalina.java:629)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:622)
	at org.apache.catalina.startup.Bootstrap.start(Bootstrap.java:289)
	at org.apache.catalina.startup.Bootstrap.main(Bootstrap.java:414)
2014-12-23 11:03:17,111 INFO  [main] org.apache.coyote.http11.Http11Protocol - Stopping Coyote HTTP/1.1 on http-8080
```

The references of com.hazelcast.util.UuidUtil$1 should be cleaned during Hazelcast shutdown.

The NPE at the very end is a bit strange, but maybe you have an idea what happens there.

Thanks and best,
Lukas

**Comment by trajano** [comment]
I am getting the same problem with Glassfish
<pre>
Severe: The web application [] created a ThreadLocal with key of type 
[com.hazelcast.util.UuidUtil$1] (value [com.hazelcast.util.UuidUtil$1@5553fb01]) 
and a value of type [java.util.Random] (value [java.util.Random@5b5032fd]) 
but failed to remove it when the web application was stopped. 
Threads are going to be renewed over time to try and avoid a probable memory leak.</pre>

**Comment by trajano** [comment]
com.hazelcast.util.UuidUtil is creating a thread local for randomizers.  Is SecureRandom() not enough for the task?

**Comment by enesakar** [comment]
closing this, in parallel to #8387

**Comment by Mario-Eis** [comment]
Will this issue be fixed in 3.7?

**Comment by pveentjer** [comment]
We are not going to fix this issue since it isn't an issue.

**Comment by Donnerbart** [comment]
This has been fixed in Hazelcast 3.8 via https://github.com/hazelcast/hazelcast/pull/9691
